### PR TITLE
feat: add update method to register endpoint

### LINKED ISSUE #60: Add modification functionality to `POST` handler for `register`
Add a `PUT` handler to the `register` endpoint to append/remove tags from existing subscriptions.

### PR DESCRIPTION
**Pull request by xav**
# Description

Implement the update functionality in the `register` endpoint
Resolves #60 

## How Has This Been Tested?

Unit tests

## Due Diligence

* [ ] Breaking change
* [ ] Requires a documentation update
* [ ] Requires a e2e/integration test update

### CODE REVIEW

File: src/handlers/register.rs [hunk 1]
```diff
@@ -32,24 +34,89 @@ pub async fn handler(
...
+        .get_registration(client_id.as_ref())
+        .await?;
+
+    let mut tags = registration.tags;
+
+    if let Some(remove_tags) = remove_tags {
+        for r_tag in remove_tags.iter() {
+            if let Some(index) = tags.iter().position(|tag| tag.eq(r_tag)) {
+                tags.remove(index);
+            }
```
Comment (chris13524) [inline src/handlers/register.rs:109]: Will this not result in a "concurrent modification exception" if using Java? I.e. removing while at the same time iterating can result in wrong indexes

File: src/handlers/register.rs [hunk 2]
```diff
@@ -19,7 +19,9 @@ use {
 #[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq)]
 #[serde(rename_all = "camelCase")]
 pub struct RegisterPayload {
-    pub tags: Vec<Arc<str>>,
+    pub tags: Option<Vec<Arc<str>>>,
+    pub append_tags: Option<Vec<Arc<str>>>,
+    pub remove_tags: Option<Vec<Arc<str>>>,
```
Comment (chris13524) [inline src/handlers/register.rs:24]:
```suggestion
    pub tags: Option<HashSet<Arc<str>>>,
    pub append_tags: Option<HashSet<Arc<str>>>,
    pub remove_tags: Option<HashSet<Arc<str>>>,
```

File: src/handlers/register.rs [hunk 3]
```diff
@@ -32,24 +34,89 @@ pub async fn handler(
...
+        }
+    }
+
+    if let Some(append_tags) = append_tags {
+        for a_tag in append_tags.iter() {
+            if !tags.contains(a_tag) {
+                tags.push(a_tag.clone());
+            }
+        }
+    }
```
Comment (chris13524) [inline src/handlers/register.rs:119]: Would a HashSet be better?

```suggestion
    let new_tags = registration.tags.into_iter().collect::<HashSet<_>>()
        .difference(&remove_tags.unwrap_or_else(|| HashSet::new())).collect::<HashSet<_>>()
        .union(&append_tags.unwrap_or_else(|| HashSet::new())).collect::<HashSet<_>>()
        .into_iter().collect::<Vec<_>>();
```

File: src/handlers/register.rs [hunk 4]
```diff
@@ -32,24 +34,89 @@ pub async fn handler(
...
+    append_tags: Option<Vec<Arc<str>>>,
+    remove_tags: Option<Vec<Arc<str>>>,
+    relay_url: Arc<str>,
+) -> error::Result<Response> {
+    if let (Some(append_tags), Some(remove_tags)) = (append_tags.clone(), remove_tags.clone()) {
+        for tag in remove_tags.iter() {
+            if append_tags.contains(tag) {
+                return Err(Error::InvalidUpdateRequest);
+            }
+        }
```
Comment (chris13524) [inline src/handlers/register.rs:95]:
```suggestion
        if remove_tags.intersect(&append_tags).count() > 0 {
            return Err(Error::InvalidUpdateRequest);
        }
```

**Review by chris13524 — CHANGES_REQUESTED** [review]

File: src/handlers/register.rs (hunk 3 above)
Comment (xav) [reply src/handlers/register.rs:119]: This one doesn't compile. A form that actually compiles is this:

```rust
    let tags = registration
        .tags
        .into_iter()
        .collect::<HashSet<_>>()
        .difference(&remove_tags.unwrap_or_else(|| HashSet::new()))
        .map(|t| t.clone())
        .collect::<HashSet<_>>()
        .union(&append_tags.unwrap_or_else(|| HashSet::new()))
        .map(|t| t.clone())
        .collect::<HashSet<_>>()
        .into_iter()
        .collect::<Vec<_>>();
```

This is really verging on write-only code, and I'm not sure there are a lot of computational advantages to this form (especially given the expected size of the collections).

Comment (chris13524) [reply src/handlers/register.rs:119]: My concern was more over readability/guaranteed set behavior than performance. Thought set operations would be easier to understand and easier to make sure the correct behavior is implemented.

Comment (chris13524) [reply src/handlers/register.rs:119]: I think this is a slightly cleaned up version that compiles:

```
    let tags = registration
        .tags
        .into_iter()
        .collect::<HashSet<_>>()
        .difference(&remove_tags.unwrap_or_else(|| HashSet::new()))
        .cloned()
        .collect::<HashSet<_>>()
        .union(&append_tags.unwrap_or_else(|| HashSet::new()))
        .cloned()
        .collect::<Vec<_>>();
```

Comment (chris13524) [reply src/handlers/register.rs:119]: Up to you. Just a suggestion

**Review by chris13524 — APPROVED** [review]

File: src/handlers/register.rs (hunk 3 above)
Comment (xav) [reply src/handlers/register.rs:119]: Sure, and I understand that in some way it's more elegant and could be more efficient on bigger collections, but I don't think we'll ever get anything much bigger than 10 items, so I'd rather go with legibility 😀

Comment (chris13524) [reply src/handlers/register.rs:119]: Again, performance wasn't my concern haha. Set operations are more legible IMO than loops and if statements

File: src/handlers/register.rs (hunk 4 above)
Comment (xav) [reply src/handlers/register.rs]: Found a middle ground 😊

Comment (xav) [reply src/handlers/register.rs]: I think the necessity to collect the iterators between each step feels kinda stupid tho, bad language design 😊

Comment (chris13524) [reply src/handlers/register.rs]: Agree, what you did is much nicer!

**Pull request merged**